### ISSUE TITLE
alignment trap on ARM

### ISSUE THREAD

**Issue by chshaob1** [msgpack/msgpack-c#460]
Hi, msgpackers:

I'm trying "zmq + msgpack-c" to make some prototype on our embedded systems. In principal, we have one device based on PowerPC, and another on ARM (armvt5). It works fine on PowerPC. However, if I send a message between two processes on ARM or send a message from PowerPC to ARM, I will see the "alignment trap" prints, like below.

 `Alignment trap: process (601) PC=0x000155d4 Instr=0xe1cc20d8 Address=0xb510450c FSR 0x001`

Then, I want to know exactly where causes the problem, so I change the ARM build from release mode (-O2) to debug mode (no optimizations at all, so nothing). Surprisingly, the "alignment trap" prints do not show up.  

I'm a bit stuck now and do not know which way to go. Any suggestions? Thanks!

**Comment by chshaob1** [comment]
The only relevant patch I could find about ARM alignment trap is [exc_bad_access in unpack_template for doubles] (https://github.com/msgpack/msgpack/issues/3)
However, I find actually current "unpack_template.h" in msgpack uses the following macro, 
which is doing the same as the patch. 
```
#define _msgpack_load64(cast, from, to) do {       \
        memcpy((cast*)(to), (from), sizeof(cast)); \
        *(to) = _msgpack_be64(*(to));              \
    } while (0);
```

any ideas?

**Comment by redboltz** [comment]
Hi @chshaob1 ,

> Then, I want to know exactly where causes the problem, so I change the ARM build from release mode (-O2) to debug mode (no optimizations at all, so nothing). Surprisingly, the "alignment trap" prints do not show up.

If you are using some data structures and memcpy binary data into the structure, the situation above might happen.
http://www.drdobbs.com/cpp/padding-and-rearranging-structure-member/240007649

I couldn't find that kinds of code on msgpack-c, so far. If you show me a small code that reproduces the problem, it would help to understand what the problem is.

**Comment by chshaob1** [comment]
Hi, redboltz,

I nailed down with the following small program. The last line matters. If I comment it out, no alignment trap. I do not know if it is easy for you to reproduce it. Probably you need an ARM processor and toolchain. Our toolchain is from poky jethro (http://git.yoctoproject.org/cgit/cgit.cgi/poky/log/?h=jethro), with gcc 5.20. Tell me if you need more information. 

`#include "msgpack.hpp"

struct TestMessage {
  int  testInt;
  bool testBool;
  std::string testString;
  MSGPACK_DEFINE(testInt, testBool, testString);
};

int main() {
  TestMessage cMsg;
  std::stringstream requestBuf;
  cMsg.testBool = false;
  cMsg.testInt = 2000;
  cMsg.testString = "from client to server process";
  msgpack::pack(requestBuf, cMsg);
  auto oh_message = msgpack::unpack(requestBuf.str().data(), requestBuf.str().size());
  msgpack::object obj = oh_message.get();
  
  obj.as<TestMessage>(); // No alignment trap shows up, if this is commented out. 
}`

**Comment by redboltz** [comment]
> I do not know if it is easy for you to reproduce it. Probably you need an ARM processor and toolchain. Our toolchain is from poky jethro with gcc 5.20. Tell me if you need more information.

It's not so easy to prepare the environment to reproduce the problem. But I can get more information by asking you.

I'd like to ask you three questions.

1 . Do you get the same error on the following code?
```C++
#include "msgpack.hpp"

struct TestMessage {
  int  testInt;
  bool testBool;
  std::string testString;
  MSGPACK_DEFINE(testInt, testBool, testString);
};

int main() {
  TestMessage cMsg;
  msgpack::zone z;
  cMsg.testBool = false;
  cMsg.testInt = 2000;
  cMsg.testString = "from client to server process";
  msgpack::object obj(cMsg, z);

  obj.as<TestMessage>(); // No alignment trap shows up, if this is commented out. 
} 
```

2 . Does the error only happen when you set `-O2` option in your minimal code?
3 . What will happen when you remove testInt, testBool, or testString, (or their combination)?

**Comment by chshaob1** [comment]
Hi, redboltz,

1. with the code you suggested, no alignment traps show up. what does this imply?

2. so far, I only tried with debug (no optimization flags, no such symptoms) and (O2, with symptoms).
I have not tried with O1, etc., since our build configuration only contain these two. 

3. I tried the following 6 combinations:  
a) int only: the struct is defined as follows. 
```
struct TestMessage {
  int  testInt;
  MSGPACK_DEFINE(testInt);
}; 
```
result: alignment trap; 

b) bool only: no alignment trap;
c) string only: no alignment trap;
d) int + bool: alignment trap; 
e) int + string: alignment trap; 
f) bool + string: no alignment trap;
The combination order is also as described above. It seems, without int, no alignment trap will show up.

**Comment by redboltz** [comment]
> 1 . with the code you suggested, no alignment traps show up. what does this imply?

I thought that the problem happens at `obj.as<TestMessage>(); `. Because you said that if you commented out the line, no alignment trap shows up. This means the problem is not caused by unpacking process, but caused by `as` (converting) process. My code doesn't contain unpacking process. I expected that you got an alignment trap if the problem is caused by `as` process. However, it didn't. It's difficult to understand. Perhaps some seeds of the problem are sown by unpacking process, and then germinate at the `at` process. Could you run `valgrind` on your environment? It might detect the seeds of the problem. Command is `valgrind ./a.out`.


> 2 . so far, I only tried with debug (no optimization flags, no such alignment symptoms) and (O2, result in such alignment symptoms). I have not tried with O1, etc., since our build configuration only contains these two.

Before digging into the compiler option, I'd like to know your compiler. If you are using `gcc` or `clang`, show me the output of `-v` option. Also I'd like to know the compiler option when you compile the code above. You've already mentioned `-O2` option. If you set other options, please let me know. Especially, I'd like to know C++ standard option. When you set `-std=c++03` or `-std=c++11`, is the alignment trap result changed?

**Comment by chshaob1** [comment]
Hi, redboltz,

It seems valgrind is not supporting armvt5 [valgrind-cross-compilation-for-armv5te](http://stackoverflow.com/questions/4228339/valgrind-cross-compilation-for-armv5tel). Just to have a try, I cheated the configure file to cross compile it, but when I start with "valgrind ls -l" to check, it reports some errors like.

```
==519== Memcheck, a memory error detector
==519== Copyright (C) 2002-2012, and GNU GPL'd, by Julian Seward et al.
==519== Using Valgrind-3.8.1 and LibVEX; rerun with -h for copyright info
==519== Command: ls -l
==519== 

vex: priv/host_arm_defs.c:2838 (imm32_to_iregNo_EXACTLY2): Assertion `0' failed.
vex storage: T total 0 bytes allocated
vex storage: P total 0 bytes allocated

valgrind: the 'impossible' happened:
   LibVEX called failure_exit().
==519==    at 0x3809DE10: ??? (in /usr/lib/valgrind/memcheck-arm-linux)

sched status:
  running_tid=1

Thread 1: status = VgTs_Runnable
==519==    at 0x46DE0AA0: ??? (in /lib/ld-2.22.so)


Note: see also the FAQ in the source distribution.
It contains workarounds to several common problems.
In particular, if Valgrind aborted or crashed after
identifying problems in your program, there's a good chance
that fixing those problems will prevent Valgrind aborting or
crashing, especially if it happened in m_mallocfree.c.

If that doesn't help, please report this bug to: www.valgrind.org

In the bug report, send all the above text, the valgrind
version, and what OS and version you are using.  Thanks.
```
apparently, my cheating does not work.

**Comment by redboltz** [comment]
Thank you for reporting. Unfortunately, it seems that we can't use valgrind. We need do more try-and-error.

First of all, could you remove `-fno-strict-aliasing`?

I think that unpacking or converting `int` is a key of the issue. So, Could you replace the type of `testInt`?
```C++
struct TestMessage {
  int  testInt;
  MSGPACK_DEFINE(testInt);
}; 
```

|check program|int8_t|int16_t|int32_t,int64_t|uint8_t|uint16_t|uint32_t,uint64_t|
|---|---|---|---|---|---|---|---|---|
|unpack and convert||||||||||
|unpack only||||||||||
|convert only||||||||||

* `unpack and convert` means your original program
* `unpack only` means remove `obj.as<TestMessage>();` from your original program.
* `convert only` means the code I wrote.

I'd like to know which case reproduce the error. It would be a hint. I'm guessing the size of integer is related to the error. I also want to know that your platform architecture ether 32bit or 64bit.

For example, `covert only` and `int8_t` as follows:

```C++
#include "msgpack.hpp"

struct TestMessage {
  int8_t  testInt;
  MSGPACK_DEFINE(testInt, testBool, testString);
};

int main() {
  TestMessage cMsg;
  msgpack::zone z;
  cMsg.testInt = 42; // need to change the value less than 128 due to int8_t.
  msgpack::object obj(cMsg, z);

  obj.as<TestMessage>(); // No alignment trap shows up, if this is commented out. 
}
```

**Comment by chshaob1** [comment]
Hi, redboltz, 

I tried all these ints, and all of them behave the same: 
1. unpack and convert: alignment trap;
2. unpack only: no alignment trap;
3. convert only: no alignment trap.

One more thing, if I pack an "int", without using struct, no alignment trap shows up, as follows.
```
int i = 3;
msgpack::pack(requestBuf, i);
auto oh_message = msgpack::unpack(requestBuf.str().data(), requestBuf.str().size());
msgpack::object obj = oh_message.get();
obj.as<int>();
```

As int8_t, and uint8_t actually uses "char", and "unsigned char", this is bit surprise to me.   

I also googled a bit about how to locate the problem from the alignment trap, and I find some interesting posts [address](http://stackoverflow.com/questions/7339489/how-to-map-pc-armv5-address-to-source-code), [resolve ARM alignment](https://gcc.gnu.org/ml/gcc/2010-05/msg00142.html), [mozilla](https://bugzilla.mozilla.org/show_bug.cgi?id=523581)
by `addr2line`, it indicates the following line: 
`**msgpack/include/msgpack/v2/object_fwd.hpp:25**`

Then, I saw the two functions on the same line, so I made the a newline, then the "addr2line" indicates line 26. It is somewhere around. 
```
struct object : v1::object {
    object() {}
    object(v1::object const& o): //line 25
      v1::object(o) {}  // line 26 
  ...
}
```
I guess there may be some casting here, which may cause unalignment access. 

By adding "printf" to trace the function calls, from "obj.as<TestMessage>()", you will find this "object" constructor function will be called 3 times. Why is it called 3 times? Also I was completely lost by different versions of "object". how is this v2::object constructor triggered? Is there any development guide somewhere? I really need some more insights. 

Another thing is that, I tried the asan from google, unfortunately, asan itself triggers more alignment trap on the my platform. So I guess, the tools that is used for general memory bugs (overflow, invalid access, etc.) will not apply here. 

I have not tried disable "-fno-strict-aliasing" yet, as the code was embedded in a large code base, which shares the same compiler flags. I may try to isolate the code out first. My platform is 32bits.

**Comment by redboltz** [comment]
> By adding "printf" to trace the function calls, from "obj.as()", you will find the above suspicious "object" constructor function will be called 3 times. Why is it called 3 times? 

v2::object inherits v1::object. v2::object has no member variables. They should be the same layouts. When I introduce v2::object, I tried to minimize the implementation for the v2::object. So I v2::object uses v1::object code. That means v1::object code treats v2::object. At that time, object converting mechanism is called. During that mechanism, v2::object is coverted from v1::object, then the constructor is called.

I believe that there is no downcast during that process even if the layout of v1::object and v2::object are the same.

I think that it is worth trying `-DMSGPACK_DEFAULT_API_VERSION=1` option. If you set the option, msgpack-c doesn't use v2 feature including v2::object.

**Comment by chshaob1** [comment]
Hi, redboltz, I'm back after quite some time. These days, I get some spare time from daily job, and take a look at the issue again. It seems I find the root cause. 

First, this is the small snipt that i use to reproduce the problem:
```
struct TestMessage_int16_t {
  int16_t  testInt;
  MSGPACK_DEFINE(testInt);
};

int main(void) {
  TestMessage_int16_t cMsg_int16_t;
  std::stringstream requestBuf;
  cMsg_int16_t.testInt = 42;
  msgpack::pack(requestBuf, cMsg_int16_t);
  auto oh_message = msgpack::unpack(requestBuf.str().data(), requestBuf.str().size());
  msgpack::object obj = oh_message.get();
  obj.as<TestMessage_int16_t>();
}
```

The root cause is at unpacking step, while execution of the parsing, the method start_array() gets called on visitor holder. Inside this method, there is a function call: 
```
obj->via.array.ptr =
                static_cast<msgpack::object*>(m_zone->allocate_align(num_elements*sizeof(msgpack::object)))
```
the problem with this call is the default argument for second parameter "align" is 4 bytes. 
```
void* allocate_align(size_t size, size_t align = MSGPACK_ZONE_ALIGN); 
#ifndef MSGPACK_ZONE_ALIGN
#define MSGPACK_ZONE_ALIGN sizeof(void*)
#endif
```
so per default it is always 4 bytes aligned. However, as in v1::object, the union type "via" uses uint64, int64, etc., which should be 8 bytes aligned. That's why whenever it is accessed, e.g., obj->via.u64 = v from visit_positive_integer(), etc., there is alignment issue, since the obj (so as is the "obj->via") is not 8 bytes aligned. 

If I add the alignment argument, e.g.,
```
obj->via.array.ptr =
                static_cast<msgpack::object*>(m_zone->allocate_align(num_elements*sizeof(msgpack::object), alignof(msgpack::object)));
```
Then it works fine. Actually this function is used quite extensively (over 160 places), so it actually creates some performance overhead, which is invisible from x86 or powerpc platform (unalignment access is handled by cpu). 

To fix it, i would propose to add the proper alignment size. 
@redboltz, what do you think? If you agree, I can prepare a patch for your review.

**Comment by redboltz** [comment]
I will write my comment on #518.

**Comment by mikeowens** [comment]
I am running into a similar issue (if not the same) on SPARC. Almost every example in the distribution results in a bus error from alignment issue on Solaris 10 SPARC . This occurs within C as well, specifically with the the object.via member.

As ARM and SPARC are both RISC and stringent with alignment, I am willing to bet this is the same issue and it therefore exists on all other RISC processors.

I can provide shell access to my server for testing/debugging purposes.

**Comment by mikeowens** [comment]
Ok I have an update. If I change the following:

```#define MSGPACK_ZONE_ALIGN sizeof(void*)```

to:

```#define MSGPACK_ZONE_ALIGN sizeof(int64_t)```

In zone.h and zone_decl.hpp, then everything works perfectly.

**Comment by redboltz** [comment]
@mikeowens , #518 has been sent by @chshaob1. It works well on travis-ci, and I guess that works well on arm environment too. But some errors are reported on appveyor (Windows). So the status of #518 is pending, so far.

The essential idea of #518 is following: 
https://github.com/msgpack/msgpack-c/pull/518/files#diff-2ccd85331cca0dc8226822ab240b36adR30

I think it is more flexible and effective than fixed 64bit setting.

@mikeowens , could you check #518 on your environment?
If it works well, I will fix windows problem.

**Comment by mikeowens** [comment]
I checked out https://github.com/msgpack/msgpack-c/pull/518/commits/b4786711dfabd9d5d17bfe0ff2113127dcee8e6a and compiled. Still bus errors:

```
./simple_c 
93 01 c3 a7 65 78 61 6d 70 6c 65 
Bus Error (core dumped)
```

**Comment by mikeowens** [comment]
The C++ examples work though.

**Comment by chshaob1** [comment]
@redboltz I'm back after reading the post from mikeowens. At my last trial, the patch breaks the appveyor (windows), which i have little clue. Then I stop. Sorry. 

you said "I think it is more flexible and effective than fixed 64bit setting." 
I just have a bit different opinion on this. In x86 or powerPC, where we do not see these alignment errors, it is because these processors take extra effort to do the corrections. 
As 8 is the right alignment size for "object.via", so why not take it?

but still, i have no idea about why the patch fails on windows. maybe there are some other issues.

**Comment by redboltz** [comment]
@mikeowens 
Thank you for checking. #518 seems to contain only C++ fix. So simple_c failed.

I think that there are three different issues.

### 1. Efficient alignment interface
I think that #518 provides type based alignment functionality. I think that #518 can achieve minimal alignment adjustment based on the type of allocating class. It is efficient. That the reason I said "I think it is more flexible and effective than fixed 64bit setting.". If we use MSGPACK_ZONE_ALIGN, all types are restricted the same conservative alignment. I think that it is not efficient. Am I missing something?
The approach depends on https://github.com/msgpack/msgpack-c/blob/master/include/msgpack/v1/detail/cpp11_zone.hpp#L230

```cpp
void* allocate_align(size_t size, size_t align = MSGPACK_ZONE_ALIGN);
```

Provides custom size alignment. And the default size is MSGPACK_ZONE_ALIGN.

### 2. How to fix C part
Currently, Provides custom size alignment functionality is in C++ only. The C part is 
https://github.com/msgpack/msgpack-c/blob/master/include/msgpack/zone.h#L106

```c
void* msgpack_zone_malloc(msgpack_zone* zone, size_t size)
```

If we have the following function, then C part will have custom alignment as C++ part.

```c
void* msgpack_zone_malloc_custom_align(msgpack_zone* zone, size_t size, size_t align)
```

### 3. How to fix alignment problem on MSVC
#518 has runtime error on msvc (appveyor).

See
https://ci.appveyor.com/project/redboltz/msgpack-c/build/2.0.0.376/job/0f7blmh87ww847kv

and search `SEH exception`.

You can see the following error:

```
)
17: [ RUN      ] object.clone_bin
17: [       OK ] object.clone_bin (0 ms)
17: [ RUN      ] object.clone_array
17: unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
17: [  FAILED  ] object.clone_array (1 ms)
17: [ RUN      ] object.clone_map
17: [       OK ] object.clone_map (0 ms)
```

The test code is https://github.com/msgpack/msgpack-c/blob/master/test/object.cpp#L393

I said "If it works well, I will fix windows problem.", but the problem is not so easy for me.

I found https://stackoverflow.com/a/28754389/1922763. I'm not sure it is related to the problem or not.

**Comment by mikeowens** [comment]
@redboltz Well I have another practical solution to offer. I added the following to the CMakeLists.txt file to detect SPARC:

```
if( CMAKE_SYSTEM_PROCESSOR STREQUAL "sparc")
    SET (CMAKE_C_FLAGS   "-DMSGPACK_ZONE_ALIGN=8 ${CMAKE_CXX_FLAGS}") 
    SET (CMAKE_CXX_FLAGS "-DMSGPACK_ZONE_ALIGN=8 ${CMAKE_CXX_FLAGS}")
endif()                                                 
```

With this, all examples (C and C++) work without issue.

Taking this route, the fix requires no code changes and only modifications to the build files and documentation. In the build files you just detect the platform and add the appropriate defines. In the documentation you warn people who are embedding the code to ensure they have the appropriate setting for alignment in their system.

This approach also has the added benefit of avoiding the Windows problems because again, there are no code changes.

**Comment by redboltz** [comment]
@mikeowens , thank you for the comment. 

I think the idea you wrote is good.

```cmake
if( CMAKE_SYSTEM_PROCESSOR STREQUAL "sparc")
    SET (CMAKE_C_FLAGS   "-DMSGPACK_ZONE_ALIGN=8 ${CMAKE_CXX_FLAGS}") 
    SET (CMAKE_CXX_FLAGS "-DMSGPACK_ZONE_ALIGN=8 ${CMAKE_CXX_FLAGS}")
endif() 
```

I found a typo. It should be as follows:

```cmake
if( CMAKE_SYSTEM_PROCESSOR STREQUAL "sparc")
    SET (CMAKE_C_FLAGS   "-DMSGPACK_ZONE_ALIGN=8 ${CMAKE_C_FLAGS}") 
    SET (CMAKE_CXX_FLAGS "-DMSGPACK_ZONE_ALIGN=8 ${CMAKE_CXX_FLAGS}")
endif() 
```

By the way, `MSGPACK_ZONE_ALIGN` is used at the following part of codes:

https://github.com/msgpack/msgpack-c/search?q=MSGPACK_ZONE_ALIGN+&type=Code&utf8=%E2%9C%93

All of them are in header files.

The C++ part of msgpack-c is a header only library. The C part have a library from `.c` file. But `MSGPACK_ZONE_ALIGN` is not used in the library.

That means the definition of `MSGPACK_ZONE_ALIGN` is for building tests and examples. The msgpack-c users need to define `MSGPACK_ZONE_ALIGN` in their project.

I will add

```cmake
if( CMAKE_SYSTEM_PROCESSOR STREQUAL "sparc")
    SET (CMAKE_C_FLAGS   "-DMSGPACK_ZONE_ALIGN=8 ${CMAKE_C_FLAGS}") 
    SET (CMAKE_CXX_FLAGS "-DMSGPACK_ZONE_ALIGN=8 ${CMAKE_CXX_FLAGS}")
endif() 
```

to  CMakeLists.txt and add document. I think https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_configure is a good place to write about `MSGPACK_ZONE_ALIGN`.

#518 is kept open for the efficient fix in the future.

**Comment by mikeowens** [comment]
Works for me. The typo was just error in cut/paste from terminal to browser. Your correction is what I had in my actual CMakeLists.txt file. Thanks for your help in this matter and with this issue.

**Comment by redboltz** [comment]
@mikeowens , I just posted #617. Could you check it on your environment?

**Comment by mikeowens** [comment]
Works perfectly.

**Comment by redboltz** [comment]
merged. @mikeowens , thank you for your cooperation.

**Comment by redboltz** [comment]
I think that I've fixed MSVC problem. It is not MSVC specific problem but zone expansion algorithm problem.
@mikeowens, could you check https://github.com/redboltz/msgpack-c/tree/chshaob1-alignment-2 on your environment? C++ part should be fixed. It is also pass appveyor tests.

The essential part of the fix is https://github.com/redboltz/msgpack-c/commit/288a6b2e315c2b89e0b9a8b6638044f50c3d7c89#diff-0c278a455197b68629ac2169cddef7d4

**Comment by mikeowens** [comment]
I checked it. Everything works fine (c and cpp). No bus errors.

**Comment by redboltz** [comment]
@mikeowens , thank you for checking! I just merged #518. It achieves type based efficient alignment adjusting on C++.  For C, accessing to the type information requires a big change in the current architecture. So I think #617 is good enough to C.
I appreciate your help.

**Comment by redboltz** [comment]
@mikeowens , you said https://github.com/msgpack/msgpack-c/issues/460#issuecomment-316260369.
How many bytes does `void*` have? 4? If it does, 32bit addressing and 64bit alignment required? I'm just curious.

**Comment by mikeowens** [comment]
According to gdb it is 4 bytes. SPARC and I think Solaris in general is capable of running both 32 and 64 bit binaries in various contexts. I am not terribly familiar with Solaris. This is a 64bit SPARC machine I bought off Ebay for $75 a couple years ago and I know just enough to compile software.

Theoretically, your approach should work correctly -- using sizeof(void*) to determine alignment. I researched it and that seemed to be the consensus I found. Why this approach doesn't work on this machine was a (frustrating) mystery to me. All I discovered was that (1) SPARC/RISC processors are unforgiving when it comes to alignment issues and (2) when I force 8 byte alignment on this machine everything worked perfectly.
